src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

```diff
@@ -311,6 +315,104 @@ internal static class PipelineOps
             return commandProcessor;
         }
 
+        private static ConditionalWeakTable<ScriptBlockAst, ScriptBlock> s_astRewriteCache = new ConditionalWeakTable<ScriptBlockAst, ScriptBlock>();
+        private static ConditionalWeakTable<ScriptBlockAst, ScriptBlock>.CreateValueCallback s_astRewriteCallback =
+            sbAst =>
+            {
+                ScriptBlockAst newScriptBlockAst = new ScriptBlockAst(
+                    sbAst.Extent,
+                    paramBlock: null,
+                    beginBlock: null,
+                    processBlock: (NamedBlockAst)sbAst.EndBlock.Copy(),
+                    endBlock: null,
+                    dynamicParamBlock: null);
+                newScriptBlockAst.PostParseChecksPerformed = sbAst.PostParseChecksPerformed;
+                sbAst.Parent?.SetParent(newScriptBlockAst);
+
+                return new ScriptBlock(newScriptBlockAst, isFilter: false);
+            };
+
+        private static bool TryRewriteForEachObjectCommand(
+            ExecutionContext context,
+            SessionStateInternal commandSessionState,
+            CommandParameterInternal[] commandElements,
+            ref CommandProcessorBase commandProcessor,
```

Comment:
Why do we need `ref` for the class parameter?

Reply:
CommandProcessorBase is a reference type and I'd expect that it is passed by reference by default.

Reply:
Thanks for clarify! Look like a trick because it is unusual for reference types. Protection comment would be good.

Reply:
OK, will add a comment here.

Reply:
I thought about what the comment should be, but the more I want to explain it clear the more I realize it become a comment that tries to explain a C# language feature ...
This is how `ref` parameter is used in C#, it passes in the variable address, and using the `ref` parameter in the method will be like directly operating on the variable from the caller method. This is essentially how local functions can directly change the variables in the enclosing methods.

Reply:
That's true, but given most analysis tools typically issue a warning for using `ref` with a reference type, a comment is probably advisable so that it's not accidentally removed. Even if that comment is "yes this is intended, don't change it" :)

Reply:
I searched the code base, there are many places `ref` is used for a reference type parameter, such as `Exception`. `Interlock.Exchange (ref object location1, object value)` also uses `ref` for reference type.
There is no difference between `ref CommandProcessorBase commandProcessor` and `ref int commandIndex` because both are aimed to directly mutate the variables in the caller method.
Personally, I don't think a comment is needed here.

Reply:
#Closed